File: plugin_execution_providers/tensorrt/example/plugin_ep_inference.py
import onnxruntime as onnxrt
import plugin_trt_ep
import numpy as np

# Path to the plugin EP library
ep_lib_path = plugin_trt_ep.get_path()
# Registration name can be anything the application chooses
ep_registration_name = "TensorRTEp"
# EP name should match the name assigned by the EP factory when creating the EP (i.e., in the implementation of OrtEP::CreateEp)
ep_name = ep_registration_name

# Register plugin EP library with ONNX Runtime
onnxrt.register_execution_provider_library(ep_registration_name, ep_lib_path)

#
# Create ORT session with explicit OrtEpDevice(s)
#

# Find the OrtEpDevice for "TensorRTEp"
ep_devices = onnxrt.get_ep_devices()
trt_ep_device = None
for ep_device in ep_devices:
    if ep_device.ep_name == ep_name:
        trt_ep_device = ep_device

assert trt_ep_device != None

sess_options = onnxrt.SessionOptions()

# Equivalent to the C API's SessionOptionsAppendExecutionProvider_V2 that appends "TensorRTEp" to ORT session option
sess_options.add_provider_for_devices([trt_ep_device], {'trt_engine_cache_enable': '1'})

assert sess_options.has_providers() == True

# Create ORT session with "TensorRTEp" plugin EP
sess = onnxrt.InferenceSession("C:\\models\\mul_1.onnx", sess_options=sess_options)

# Run sample model and check output
x = np.array([[1.0, 2.0], [3.0, 4.0], [5.0, 6.0]], dtype=np.float32)
input_name = sess.get_inputs()[0].name
res = sess.run([], {input_name: x})
output_expected = np.array([[1.0, 4.0], [9.0, 16.0], [25.0, 36.0]], dtype=np.float32)
np.testing.assert_allclose(output_expected, res[0], rtol=1e-05, atol=1e-08)

# Unregister the library using the application-specified registration name.
# Must only unregister a library after all sessions that use the library have been released.
onnxrt.unregister_execution_provider_library(ep_registration_name)


# Note:
# The mul_1.onnx can be found here:
# https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/test/testdata/mul_1.onnx

File: c_cxx/plugin_EP/plugin_ep_inference.cc
#include "onnxruntime_cxx_api.h"
#include <iostream>
#include <vector>
#include <gsl/gsl>

int RunInference() { 
  const OrtApi* ort_api = OrtGetApiBase()->GetApi(ORT_API_VERSION);
  Ort::Env env;

  // Registration name can be anything the application chooses
  const char* lib_registration_name = "TensorRTEp";

  // Register plugin EP library with ONNX Runtime.
  env.RegisterExecutionProviderLibrary(
      lib_registration_name,      // Registration name can be anything the application chooses.
      ORT_TSTR("TensorRTEp.dll")  // Path to the plugin EP library.
  );

  // Unregister the library using the application-specified registration name.
  // Must only unregister a library after all sessions that use the library have been released.
  auto unregister_plugin_eps_at_scope_exit = gsl::finally([&]() { 
    env.UnregisterExecutionProviderLibrary(lib_registration_name);
  });

  {
    std::vector<Ort::ConstEpDevice> ep_devices = env.GetEpDevices();
    // EP name should match the name assigned by the EP factory when creating the EP (i.e., in the implementation of OrtEP::CreateEp())
    std::string ep_name = lib_registration_name;

    // Find the Ort::EpDevice for "TensorRTEp".
    std::vector<Ort::ConstEpDevice> selected_ep_devices = {};
    for (Ort::ConstEpDevice ep_device : ep_devices) {
      if (std::string(ep_device.EpName()) == ep_name) {
        selected_ep_devices.push_back(ep_device);
        break;
      }
    }

    if (selected_ep_devices[0] == nullptr) {
      // Did not find EP. Report application error ...
      std::cerr << "Did not find EP: " << ep_name << std::endl;
      return -1;
    }

    std::unordered_map<std::string, std::string> ep_options;  // Optional EP options.
    Ort::SessionOptions session_options;
    session_options.AppendExecutionProvider_V2(env, selected_ep_devices, ep_options);

    Ort::Session session(env, ORT_TSTR("mul_1.onnx"), session_options);

    // Get default ORT allocator
    Ort::AllocatorWithDefaultOptions allocator;

    // Get input name
    Ort::AllocatedStringPtr input_name_ptr = session.GetInputNameAllocated(0, allocator); // Keep the smart pointer alive to avoid dangling pointer
    const char* input_name = input_name_ptr.get();

    // Input data
    std::vector<float> input_values = {1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f};

    // Input shape: (3, 2)
    std::vector<int64_t> input_shape{3, 2};

    // Create tensor
    Ort::MemoryInfo memory_info = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);

    Ort::Value input_tensor = Ort::Value::CreateTensor<float>(memory_info, input_values.data(), input_values.size(),
                                                              input_shape.data(), input_shape.size());

    // Get output name
    Ort::AllocatedStringPtr output_name_ptr =
        session.GetOutputNameAllocated(0, allocator);  // Keep the smart pointer alive to avoid dangling pointer
    const char* output_name = output_name_ptr.get();

    // Run session
    std::vector<const char*> input_names{input_name};
    std::vector<const char*> output_names{output_name};

    auto output_tensors =
        session.Run(Ort::RunOptions{nullptr}, input_names.data(), &input_tensor, 1, output_names.data(), 1);

    // Extract output
    float* output_data = output_tensors.front().GetTensorMutableData<float>();

    std::cout << "Output:" << std::endl;
    for (int i = 0; i < 6; i++) {
      std::cout << output_data[i] << " ";
    }
    std::cout << std::endl;

    // Expected output: [[1,4],[9,16],[25,36]]
  }

  return 0;
}

int main(int argc, char* argv[]) {
  return RunInference();
}

// Note:
// The mul_1.onnx can be found here:
// https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/test/testdata/mul_1.onnx
